src/Build/BackEnd/Components/RequestBuilder/TargetEntry.cs

```diff
@@ -624,24 +625,34 @@ internal async Task ExecuteTarget(ITaskBuilder taskBuilder, BuildRequestEntry re
                         {
                             foreach (ItemBucket bucket in batchingBuckets)
```

Comment:
Is there a way to loop through the buckets and get a count to pass to the list ctor or directly allocate a right-sized array or HashSet? That would help all codepaths here and simplify the code, too.

Reply:
I believe that was something jeffkl and I investigated, but bucket.Expander.ExpandIntoTaskItemsLeaveEscaped returns an unknown number of items. Either we immediately add those to an object combining the results (as we currently do) or we have to throw away the results and recalculate them later, which would be more expensive. We could also have an object that retains a pointer to each of the expanded item lists, but that would require holding onto all those lists for the duration of the for loop, which is tough on gc.